### PR TITLE
[KED-1312] Add a GeoJSONDataset to kedro.extras.datasets

### LINKED ISSUE #189: [KED-1312] New DataSet : GeoJSONLocalDataSet
## Description
Create a GeoJSONLocalDataSet that would be read as a [geopandas.GeoDataFrame](http://geopandas.org/reference/geopandas.GeoDataFrame.html#geopandas.GeoDataFrame)

## Context
GeoDataFrame allow for fast spatial joins in pandas (with the use of Rtree) 

## Possible Implementation
Inherhit from [AbstractDataSet](https://github.com/lblanche/kedro/blob/bfd802db5cb72841730f5e2e1a787255658cd89d/kedro/io/core.py#L86)  and create class in contrib.io

```python
"""GeoJSONLocalDataSet loads and saves data to a local geojson file. The
underlying functionality is supported by geopandas, so it supports all
allowed geopandas (pandas) options for loading and saving geosjon files.
"""
from os.path import isfile
from typing import Any, Union, Dict

import geopandas as gpd

from kedro.io import AbstractDataSet


class GeoJSONLocalDataSet(AbstractDataSet):
    """``GeoJSONLocalDataSet`` loads and saves data to a local geojson file. The
    underlying functionality is supported by geopandas, so it supports all
    allowed geopandas (pandas) options for loading and saving geosjon files.

    Example:
    ::

        >>> import geopandas as gpd
        >>> from shapely.geometry import Point
        >>>
        >>> data = gpd.GeoDataFrame({'col1': [1, 2], 'col2': [4, 5],
        >>>                      'col3': [5, 6]}, geometry=[Point(1,1), Point(2,4)])
        >>> data_set = GeoJSONLocalDataSet(filepath="test.geojson",
        >>>                                save_args=None)
        >>>                                 
        >>> data_set.save(data)
        >>> reloaded = data_set.load()
        >>>
        >>> assert data.equals(reloaded)

    """

    def _describe(self) -> Dict[str, Any]:
        return dict(
            filepath=self._filepath,
            driver=self._driver,
            load_args=self._load_args,
            save_args=self._save_args,
        )

    def __init__(
        self,
        filepath: str,
        driver: str = "GeoJSON",
        load_args: Dict[str, Any] = None,
        save_args: Dict[str, Any] = None,
    ) -> None:
        """Creates a new instance of ``GeoJSONLocalDataSet`` pointing to a concrete
        filepath.

        Args:
            driver: The driver used to write to GeoJSON files. The default
                          engine is 'GeoJSON'.

            filepath: path to a geojson file.

            load_args: GeoPandas options for loading GeoJSON files.
                Here you can find all available arguments:

            save_args: GeoPandas options for saving geojson files.
                Here you can find all available arguments:
                The default_save_arg driver is 'GeoJSON', all others preserved.

        """
        self._filepath = filepath
        default_save_args = {"driver": "GeoJSON"}
        default_load_args = {}

        self._load_args = (
            {**default_load_args, **load_args}
            if load_args is not None
            else default_load_args
        )
        self._save_args = (
            {**default_save_args, **save_args}
            if save_args is not None
            else default_save_args
        )
        self._driver = driver

    def _load(self) -> Union[gpd.GeoDataFrame, Dict[str, gpd.GeoDataFrame]]:
        return gpd.read_file(self._filepath, **self._load_args)

    def _save(self, data: gpd.GeoDataFrame) -> None:
        data.to_file(**self._save_args)

    def _exists(self) -> bool:
        return isfile(self._filepath)

```

Remark : other types of  geo data format could be added later (eg. .shp, .kml etc)

### PR DESCRIPTION
**Pull request by lblanche**
Resolves #189 
## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

## Motivation and Context
Why was this PR created?
To be able to use geoDataFrame within kedro

## How has this been tested?
What testing strategies have you used?
Same as the CSVLocalDataSet

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](/RELEASE.md) file
- [x] Added tests to cover my changes
- [ ] Assigned myself to the PR

### CODE REVIEW

**Comment by lblanche** [comment]
I did not know of Draft Pull Requests , Ill do it that way next time ! Also I cannot assign myself  to the PR

**Comment by 921kiyo** [comment]
@lblanche No problem! Thank you for your contribution. You might have noticed in the upcoming release note, but we are pushing datasets toward more platform agnostic (e.g [JSONDataSet](https://github.com/quantumblacklabs/kedro/blob/develop/kedro/io/json_dataset.py) instead of JSONLocalDataSet etc) at the moment using `fsspec` (https://filesystem-spec.readthedocs.io/en/latest/ ). It would be great if you make your dataset platform agnostic.  
You can have a look at existing datasets. If you need help, we are more than happy to help (@andrii-ivaniuk has been working on this in the past couple of months.) :)

**Comment by lblanche** [comment]
@921kiyo I did miss that, I'll look into it when I have time then. Thanks

**Comment by lblanche** [comment]
I think this is ready  :)

**Comment by ZainPatelQB** [comment]
> I think this is ready :)

Sorry for the late reply 😅 Quick update: we'll get on with reviewing this soon. We're currently doing some work with restructuring the way `contrib` works, so merging this in will probably wait on that and we'll take care of moving your contributions to the new structure. Thank you for your contribution and for adding `fsspec` support to your dataset! 😀

**Comment by ZainPatelQB** [comment]
Does geopandas need to be pinned to 0.6.0 or can we be more liberal with the requirements?

**Comment by ZainPatelQB** [comment]
For reviewers, my changes: https://github.com/quantumblacklabs/kedro/pull/190/files/30b0154db2bd5dce95ce937ec0750c6345d60e69..d0f839b9842cc19850a539b249f89c4f984be8d0

**Comment by lblanche** [comment]
@ZainPatelQB I think we can be more liberal regarding geopandas version, I haven't tested it though

File: RELEASE.md [hunk 1]
```diff
@@ -98,7 +100,7 @@
 * `kedro.contrib.io.bioinformatics.sequence_dataset.py` was renamed to `kedro.contrib.io.bioinformatics.biosequence_local_dataset.py`.
 
 ## Thanks for supporting contributions
-[Jonas Kemper](https://github.com/jonasrk), [Yuhao Zhu](https://github.com/yhzqb), [Balazs Konig](https://github.com/BalazsKonigQB), [Pedro Abreu](https://github.com/PedroAbreuQB), [Tam-Sanh Nguyen](https://github.com/tamsanh), [Peter Zhao](https://github.com/zxpeter), [Deepyaman Datta](https://github.com/deepyaman), [Florian Roessler](https://github.com/fdroessler/)
+[Jonas Kemper](https://github.com/jonasrk), [Yuhao Zhu](https://github.com/yhzqb), [Balazs Konig](https://github.com/BalazsKonigQB), [Pedro Abreu](https://github.com/PedroAbreuQB), [Tam-Sanh Nguyen](https://github.com/tamsanh), [Peter Zhao](https://github.com/zxpeter), [Deepyaman Datta](https://github.com/deepyaman), [Florian Roessler](https://github.com/fdroessler/), [Luis Blanche](https://github/LuisBlanche)
```
Comment (limdauto) [inline RELEASE.md]:
```suggestion
[Jonas Kemper](https://github.com/jonasrk), [Yuhao Zhu](https://github.com/yhzqb), [Balazs Konig](https://github.com/BalazsKonigQB), [Pedro Abreu](https://github.com/PedroAbreuQB), [Tam-Sanh Nguyen](https://github.com/tamsanh), [Peter Zhao](https://github.com/zxpeter), [Deepyaman Datta](https://github.com/deepyaman), [Florian Roessler](https://github.com/fdroessler/), [Luis Blanche](https://github.com/LuisBlanche)
```

File: kedro/extras/datasets/geojson/geojson_dataset.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    (eg: local, S3, GCS).
+    The underlying functionality is supported by geopandas, so it supports all
+    allowed geopandas (pandas) options for loading and saving geosjon files.
+
+    Example:
+    ::
+
+        >>> import geopandas as gpd
+        >>> from shapely.geometry import Point
+        >>> from kedro.io.contrib import GeoJSONDataSet
```
Comment (921kiyo) [inline kedro/extras/datasets/geojson/geojson_dataset.py]:
```suggestion
        >>> from kedro.extras.datasets.geopandas import GeoJSONDataSet
```

File: kedro/extras/datasets/geojson/geojson_dataset.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        >>>                                save_args=None)
+        >>> data_set.save(data)
+        >>> reloaded = data_set.load()
+        >>>
+        >>> assert data.equals(reloaded)
+
+    """
+
+    DEFAULT_LOAD_ARGS = {}  # type: Dict[str, Any]
+    DEFAULT_SAVE_ARGS = {"driver": "GeoJSON"}  # type: Dict[str, Any]
```
Comment (921kiyo) [inline kedro/extras/datasets/geojson/geojson_dataset.py]: `# type: Dict[str, Any]` is probably not necessary.

File: kedro/extras/datasets/geojson/geojson_dataset.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+# license to the QuantumBlack Trademarks. You may not use the QuantumBlack
+# Trademarks or any confusingly similar mark as a trademark for your product,
+#     or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""GeoJSONDataSet loads and saves data to a local geojson file. The
```
Comment (921kiyo) [inline kedro/extras/datasets/geojson/geojson_dataset.py]:
```suggestion
"""GeoJSONDataSet loads and saves data to a local GeoJSON file. The
```

File: kedro/extras/datasets/geojson/geojson_dataset.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    AbstractVersionedDataSet,
+    DataSetError,
+    Version,
+    get_filepath_str,
+    get_protocol_and_path,
+)
+
+
+class GeoJSONDataSet(AbstractVersionedDataSet):
+    """``GeoJSONDataSet`` loads and saves data to a geojson file using an underlying filesystem
```
Comment (921kiyo) [inline kedro/extras/datasets/geojson/geojson_dataset.py]:
```suggestion
    """``GeoJSONDataSet`` loads and saves data to a GeoJSON file using an underlying filesystem
```

File: kedro/extras/datasets/geojson/geojson_dataset.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            save_args: GeoPandas options for saving geojson files.
+                Here you can find all available arguments:
+                The default_save_arg driver is 'GeoJSON', all others preserved.
+            version: If specified, should be an instance of
+                ``kedro.io.core.Version``. If its ``load`` attribute is
+                None, the latest version will be loaded. If its ``save``
+            credentials: credentials required to access the underlying filesystem.
+                Eg. for ``GCFileSystem`` it would look like `{'token': None}`.
+            fs_args: Extra args to pass into the underlying filesystem
+                Eg. for ``GCFileSystem`` it would look like `{'project': 'my-project' , ...}`.
```
Comment (921kiyo) [inline kedro/extras/datasets/geojson/geojson_dataset.py]: Can you please look at `json_dataset.py` and make the docstring more less consistent with it?

**Comment by ZainPatelQB** [comment]
@lblanche do you mind having a look at the geopandas requirements and seeing if we can loosen it to play nice with out libraries? :)

File: tests/extras/datasets/geojson/test_geojson_dataset.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+import geopandas as gpd
+import pytest
+from fsspec.implementations.http import HTTPFileSystem
+from fsspec.implementations.local import LocalFileSystem
+from gcsfs import GCSFileSystem
+from pandas.util.testing import assert_frame_equal
+from s3fs import S3FileSystem
+from shapely.geometry import Point
+
+from kedro.extras.datasets.geojson import GeoJSONDataSet
```
Comment (921kiyo) [inline tests/extras/datasets/geojson/test_geojson_dataset.py]:
```suggestion
from kedro.extras.datasets.geopandas import GeoJSONDataSet
```

File: docs/source/05_api_docs/kedro.extras.datasets.rst [hunk 8]
```diff
@@ -21,6 +21,7 @@ Data Sets
     kedro.extras.datasets.pandas.ExcelDataSet
     kedro.extras.datasets.pandas.FeatherDataSet
     kedro.extras.datasets.pandas.GBQTableDataSet
+    kedro.extras.datasets.geojson.GeoJSONDataSet
```
[the remark this answers is not in the document]
Comment (921kiyo) [reply docs/source/05_api_docs/kedro.extras.datasets.rst]: If you look at other extra datasets (e,g JSONDataSet) the path is `kedro.extras.datasets.pandas.XXX`, not json.

File: RELEASE.md [hunk 9]
```diff
@@ -1,6 +1,7 @@
 # Upcoming Release
 
 ## Major features and improvements
+* Added a `GeoJSONDataSet`dataset in `kedro.extras.datasets.geojson` for working with geospatial data that uses [`fsspec`](https://filesystem-spec.readthedocs.io/en/latest/) to communicate with the underlying filesystem.
```
Comment (921kiyo) [inline RELEASE.md]:
```suggestion
* Added a `GeoJSONDataSet`dataset in `kedro.extras.datasets.geopandas` for working with geospatial data that uses [`fsspec`](https://filesystem-spec.readthedocs.io/en/latest/) to communicate with the underlying filesystem.
```

File: docs/source/05_api_docs/kedro.extras.datasets.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (ZainPatelQB) [reply docs/source/05_api_docs/kedro.extras.datasets.rst]: If we're renaming this, can you also rename "geojson" -> "geopandas" in `extras_require` in `setup.py`? :)

File: kedro/extras/datasets/geojson/geojson_dataset.py (hunk 3 above)
Comment (mzjp2) [reply kedro/extras/datasets/geojson/geojson_dataset.py]: Most other dataset seems to have this, unless I'm mistaken?

Comment (921kiyo) [reply kedro/extras/datasets/geojson/geojson_dataset.py]: When the dict is not empty ?

Comment (mzjp2) [reply kedro/extras/datasets/geojson/geojson_dataset.py]: But then the type inferred is `Dict[str, str]` by mypy, are we sure all the values are `str`s?

**Comment by mzjp2** [comment]
All merged in. Congrats and thanks @lblanche 🎉

**Pull request merged**